src/test/auxiliary/issue_8401.rs

```diff
@@ -22,5 +22,5 @@ fn bar<T>(_: &mut A, _: &T) {}
 
 fn foo<T>(t: &T) {
     let b = B;
-    bar(unsafe { mem::transmute(&b as &A) }, t)
+    bar(&mut b as &mut A, t)
```

Comment:
I pushed before running (the run caught what you noticed), since I needed to rebuild and it'd take a while. It's all done, and I'm pushing the corrections that make these 2 tests pass.